lightning/src/util/events.rs

```diff
@@ -150,6 +150,23 @@ pub enum Event {
 		/// The outputs which you should store as spendable by you.
 		outputs: Vec<SpendableOutputDescriptor>,
 	},
+	/// This event is generated when a payment has been successfully forwarded through us and a
+	/// forwarding fee earned.
+	PaymentForwarded {
+		/// The fee, in milli-satoshis, which was earned as a result of the payment.
+		///
+		/// Note that if we force-closed a channel while a forwarded HTLC was pending, the amount
+		/// the next hop claimed may have been rounded down, causing a higher fee than expected.
+		///
+		/// If the channel which sent us the payment has been force-closed, we will claim the funds
+		/// via an on-chain transaction. In that case we do not consider the on-chain transaction
+		/// fees involved in such a claim and this value will be `None`. It is possible duplicate
+		/// `PaymentForwarded` events are generated for the same payment iff `fee_earned_msat` is
+		/// `None`.
```

Comment:
Because duplicates may be generated, could a (future?) direction be to add some kind of unique identifier of the HTLC to this event?

Reply:
Yea, what we'd really need to do is have the event be generated by the `ChannelMonitor` if the inbound leg of the payment has gone on-chain, but there's a lot of complexity to ensuring we de-duplicate them all properly that we don't currently do that I figured it wasn't worth it for now.

Reply:
Yeah we could return either a monotonic payment_id or `PaymentHash` at least.

Reply:
>  or `PaymentHash` at least.

I'm a little torn on this. I initially didn't include it on purpose because, at least my thinking was, we really dont want users to log that or even care about it - its not their payment, and logging/giving users details about the payments others made (at least the parts that don't matter to you) is generally a bad idea.

On the other hand, I see the point about wanting to de-dup by them, but of course note that you can have the same `PaymentHash` multiple times, even if its insecure.

Reply:
Right, we already have case where accounting might not be exact like rouding down or backward channel closed and where `fee_earned_msat` is `None`. We can revisit in the future if needed.